// idyll-compiler/dist/cjs/ast.js
'use strict';

/**
 *
 * This module contains utility functions
 * for dealing with Idyll's internal representation
 * of an AST. The ast object is an array of nodes,
 * where each node is defined like:
 *
 * [ name, [
 *      [property1, [valueType, value ]],
*       [property2, [valueType, value ]]
 *    ], [
 *  child1,
 *  child2,
 *  ...
 * ] ]
 *
*/

module.exports = {

  getChildren: function getChildren(node) {
    return node[2] || [];
  },

  modifyChildren: function modifyChildren(node, modifier) {
    if (typeof node === 'string') {
      return node;
    }
    node[2] = this.getChildren(node).map(function (child) {
      return modifier(child);
    });
    return node;
  },

  modifyNodesByName: function modifyNodesByName(ast, name, modifier) {
    var _this = this;

    var handleNode = function handleNode(node) {
      if (typeof node === 'string') {
        return node;
      }
      if (node[0].toLowerCase() === name.toLowerCase()) {
        node = modifier(node);
      }

      node = _this.modifyChildren(node, handleNode);
      return node;
    };

    ast = ast.map(function (node) {
      return handleNode(node);
    });
    return ast;
  },

  getProperty: function getProperty(node, key) {
    node[1].forEach(function (element) {
      if (element[0] === key) {
        return element[1];
      }
    });
  },

  setProperty: function setProperty(node, key, value) {
    var hasSet = false;
    node[1].forEach(function (element) {
      if (element[0] === key) {
        hasSet = true;
      }
    });
    if (!hasSet) {
      node[1] = node[1].concat([[key, value]]);
    }

    return node;
  },

  setProperties: function setProperties(node, properties) {
    var _this2 = this;

    Object.keys(properties).forEach(function (key) {
      node = _this2.setProperty(node, key, properties[key]);
    });
    return node;
  },

  removeProperty: function removeProperty(node, key) {
    node[1] = node[1].filter(function (_ref) {
      var propName = _ref[0],
          propVale = _ref[1];

      if (propName === key) {
        return false;
      }
      return true;
    });
    return node;
  }
};
